### PR TITLE
Add T-State factory

### PR DESCRIPTION
**Pull request by NoureldinYosri**
- Adds a simple 15-to-1 T-state factory class that extends the `MagicStateFactory` class
- Adds a new method `spacetime_footprint` to the `MagicStateFactory` which should return the spacetime consumed per single magic state created. This is the quantity used to compare and pick a factory in https://arxiv.org/abs/2211.07629

part of https://github.com/quantumlib/Qualtran/issues/269

### CODE REVIEW

File: qualtran/surface_code/ccz2t_cost_model_test.py [hunk 1]
```diff
@@ -30,3 +30,8 @@ def test_vs_spreadsheet():
     np.testing.assert_allclose(re.failure_prob, 0.0084, rtol=1e-3)
     np.testing.assert_allclose(re.footprint, 4.00e5, rtol=1e-3)
     np.testing.assert_allclose(re.duration_hr, 7.53, rtol=1e-3)
+
+
+def test_factory():
+    factory = CCZ2TFactory()
+    assert factory.spacetime_footprint() is NotImplemented
```
Comment (mpharrigan) [inline qualtran/surface_code/ccz2t_cost_model_test.py:37]: ?? this doesn't seem like a useful test. Presumably we'd want to implement this method soon, no?

File: qualtran/surface_code/magic_state_factory.py [hunk 2]
```diff
@@ -36,3 +36,7 @@ def n_cycles(self, n_magic: AlgorithmSummary) -> int:
     @abc.abstractmethod
     def distillation_error(self, n_magic: AlgorithmSummary, phys_err: float) -> float:
         """The total error expected from distilling magic states with a given physical error rate."""
+
+    @abc.abstractmethod
+    def spacetime_footprint(self) -> float:
+        """The spacetime cost of producing a single magic state."""
```
Comment (mpharrigan) [inline qualtran/surface_code/magic_state_factory.py:42]: I think this needs more documentation. What are the units? logical or physical? There are existing properties named `footprint` and `n_cycles`. Is this just the product of those two? 

Should this be derived from those other two methods? What happens if they're inconsistent?

File: qualtran/surface_code/t_factory.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+from qualtran.surface_code.algorithm_summary import AlgorithmSummary
+from qualtran.surface_code.magic_state_factory import MagicStateFactory
+
+
+@frozen
+class TFactory(MagicStateFactory):
+    """A magic state factory for T states.
+
+    Attributes:
+        num_qubits: Number of qubits used by the factory.
```
Comment (mpharrigan) [inline qualtran/surface_code/t_factory.py:28]: physical or logical?

File: qualtran/surface_code/t_factory.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+from attrs import field, frozen
+
+from qualtran.surface_code.algorithm_summary import AlgorithmSummary
+from qualtran.surface_code.magic_state_factory import MagicStateFactory
+
+
+@frozen
+class TFactory(MagicStateFactory):
+    """A magic state factory for T states.
```
Comment (mpharrigan) [inline qualtran/surface_code/t_factory.py:25]: How does this relate to the existing `CCZ2TFactory(MagicStateFactory)`? That one is a very specific construction from a particular paper. Is this supposed to be just a generic attributes-based T factory?

File: qualtran/surface_code/t_factory.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    Attributes:
+        num_qubits: Number of qubits used by the factory.
+        duration: Time taken by the factory to produce T states.
+        t_states_rate: Number of T states per production cycle.
+        reference: Source of these estimates.
```
Comment (mpharrigan) [inline qualtran/surface_code/t_factory.py:31]: these don't match the actual fields

File: qualtran/surface_code/t_factory.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    def footprint(self) -> int:
+        return self.num_qubits
+
+    def n_cycles(self, n_magic: AlgorithmSummary) -> int:
+        t_states = n_magic.t_gates + 4 * n_magic.toffoli_gates
+        return math.ceil(t_states / self.num_t_per_cycle)
+
+    def distillation_error(self, n_magic: AlgorithmSummary, phys_err: float) -> float:
+        return NotImplemented
```
Comment (mpharrigan) [inline qualtran/surface_code/t_factory.py:48]: I think this should `raise NotImplementedError`, no?

File: qualtran/surface_code/t_factory.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    Attributes:
+        num_qubits: Number of qubits used by the factory.
+        duration: Time taken by the factory to produce T states.
+        t_states_rate: Number of T states per production cycle.
+        reference: Source of these estimates.
+    """
+
+    num_qubits: int
+    generation_cycle_duration_ns: float = field(repr=lambda x: f'{x:g}')
```
Comment (mpharrigan) [inline qualtran/surface_code/t_factory.py:35]: I would strongly advocate for not mixing physical units into this definition. Isn't that the point of your other class to contain the physical attributes of a device/scheme?

**Review by mpharrigan — CHANGES_REQUESTED** [review]

File: qualtran/surface_code/ccz2t_cost_model_test.py (hunk 1 above)
Comment (NoureldinYosri) [reply qualtran/surface_code/ccz2t_cost_model_test.py:37]: for coverage

Comment (mpharrigan) [reply qualtran/surface_code/ccz2t_cost_model_test.py:37]: The idiomatic solution in cirq is to do `# coverage: skip`. Here, we don't have a coverage check so you don't need to do anything

File: qualtran/surface_code/t_factory.py (hunk 4 above)
Comment (NoureldinYosri) [reply qualtran/surface_code/t_factory.py]: The CCZ factory is as you said describies the implementation. This class is a simple data class description

File: qualtran/surface_code/t_factory.py (hunk 5 above)
Comment (NoureldinYosri) [reply qualtran/surface_code/t_factory.py]: fixed

**Review by NoureldinYosri — COMMENTED** [review]
@mpharrigan I renamed the fields to better reflect what they do. none of them is a phsical parameter.

File: qualtran/surface_code/t_factory.py (hunk 7 above)
Comment (NoureldinYosri) [reply qualtran/surface_code/t_factory.py]: ptal. I renamed the parameter to better reflect what they are. none of them are physical

File: qualtran/surface_code/ccz2t_cost_model.py [hunk 8]
```diff
@@ -150,17 +149,20 @@ def n_cycles(self, n_magic: AlgorithmSummary) -> int:
 
         return math.ceil((n_ccz_states * ccz_depth + catalyzations) * distillation_d)
 
+    def spacetime_footprint(self) -> float:
+        raise NotImplementedError
```
Comment (mpharrigan) [inline qualtran/surface_code/ccz2t_cost_model.py:153]: ()

File: qualtran/surface_code/magic_state_factory.py (hunk 2 above)
Comment (mpharrigan) [reply qualtran/surface_code/magic_state_factory.py]: bump

File: qualtran/surface_code/t_factory.py (hunk 3 above)
Comment (mpharrigan) [reply qualtran/surface_code/t_factory.py]: In another comment you say

> none of them is a phsical parameter.

?

Comment (NoureldinYosri) [reply qualtran/surface_code/t_factory.py]: well it's not a physical parameter as in time or error. It's just the number of qubits this factory. The factory described in the paper works with physical qubits

File: qualtran/surface_code/t_factory.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+@frozen
+class SimpleTFactory(MagicStateFactory):
+    """A summary of properties of a T-states factory.
+
+    A summary of T-state factories as described in Appendix E of https://arxiv.org/abs/2211.07629
+
+    Attributes:
+        num_qubits: Number of physical qubits used by the factory.
+        generation_time_ns: Time to generate a single T-state.
+        distillation_error_: Probability of not accepting a magic state
```
Comment (mpharrigan) [inline qualtran/surface_code/t_factory.py:32]: why is there a trailing underscore?

File: qualtran/surface_code/ccz2t_cost_model.py [hunk 10]
```diff
@@ -150,17 +149,20 @@ def n_cycles(self, n_magic: AlgorithmSummary) -> int:
...
     phys_err: float = 1e-3,
-    error_budget: Optional[float] = 1e-2,
+    error_budget: float = 1e-2,
     cycle_time_us: float = 1.0,
-    routing_overhead: Optional[float] = 0.5,
-    factory: MagicStateFactory = None,
-    data_block: DataBlock = None,
+    routing_overhead: float = 0.5,
+    factory: MagicStateFactory | None = None,
+    data_block: DataBlock | None = None,
```
Comment (mpharrigan) [inline qualtran/surface_code/ccz2t_cost_model.py:165]: #522

File: qualtran/surface_code/ccz2t_cost_model.py (hunk 8 above)
Comment (mpharrigan) [reply qualtran/surface_code/ccz2t_cost_model.py:153]: actually... why is this not implemented?

File: qualtran/surface_code/magic_state_factory.py (hunk 2 above)
Comment (NoureldinYosri) [reply qualtran/surface_code/magic_state_factory.py]: done

File: qualtran/surface_code/ccz2t_cost_model.py (hunk 8 above)
Comment (NoureldinYosri) [reply qualtran/surface_code/ccz2t_cost_model.py]: done

File: qualtran/surface_code/magic_state_factory.py [hunk 11]
```diff
@@ -36,3 +36,11 @@ def n_cycles(self, n_magic: AlgorithmSummary) -> int:
...
     def distillation_error(self, n_magic: AlgorithmSummary, phys_err: float) -> float:
         """The total error expected from distilling magic states with a given physical error rate."""
+
+    @abc.abstractmethod
+    def spacetime_footprint(self) -> float:
+        """The spacetime cost of producing a single magic state.
+
+        Equals the number of qubits times the time needed to produce a single magic state.
+        reference: Appendix E of https://arxiv.org/abs/2211.07629
+        """
```
[the remark this answers is not in the document]
Comment (NoureldinYosri) [reply qualtran/surface_code/magic_state_factory.py:49]: this is the definition in E4 of the azure paper. Do we wanna stick to that definition?

**Comment by NoureldinYosri** [comment]
on another thought I don't think we need this atm. We should implement a proper T factory representation at some point. but for the reproduction of the azure model we don't need this

**Pull request closed without merging**